pkg/mcs/router/server/sync.go

```diff
@@ -0,0 +1,367 @@
+	// false means reconnect only when the current streaming is not running
+	reconnectCh       chan bool
+	checkMembershipCh chan struct{}
+	etcdClient        *clientv3.Client
+	pdLeaderAddr      atomic.Value
+}
+
+// NewRegionSyncer returns a region syncer.
+func NewRegionSyncer(leaderCtx context.Context, cluster *core.BasicCluster, etcdClient *clientv3.Client,
+	tlsConfig *grpcutil.TLSConfig, name, listenURL string) *RegionSyncer {
+	checkMembershipCh := make(chan struct{}, 1)
+	reconnectCh := make(chan bool, 1)
+	return &RegionSyncer{
+		cluster:           cluster,
+		tlsConfig:         tlsConfig,
+		name:              name,
+		listenURL:         listenURL,
+		leaderCtx:         leaderCtx,
+		checkMembershipCh: checkMembershipCh,
+		etcdClient:        etcdClient,
+		reconnectCh:       reconnectCh,
+	}
+}
+
+const (
+	keepaliveTime        = 10 * time.Second
+	keepaliveTimeout     = 3 * time.Second
+	msgSize              = 8 * units.MiB
+	retryInterval        = time.Second
+	memberUpdateInterval = time.Minute
+)
+
+func (s *RegionSyncer) getClient() *clientv3.Client {
+	return s.etcdClient
+}
+
+func (s *RegionSyncer) updatePDMemberLoop() {
+	s.wg.Add(1)
+	defer logutil.LogPanic()
+	defer s.wg.Done()
+
+	ctx, cancel := context.WithCancel(s.leaderCtx)
+	defer cancel()
+	ticker := time.NewTicker(memberUpdateInterval)
+	failpoint.Inject("speedUpMemberLoop", func() {
+		ticker.Reset(100 * time.Millisecond)
+	})
+	defer ticker.Stop()
+	var curLeader uint64
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("server is closed, exit update member loop")
+			return
+		case <-ticker.C:
+		case <-s.checkMembershipCh:
+		}
+		members, err := etcdutil.ListEtcdMembers(ctx, s.getClient())
+		if err != nil {
+			log.Warn("failed to list members", errs.ZapError(err))
+			continue
+		}
+		for _, ep := range members.Members {
+			if len(ep.GetClientURLs()) == 0 { // This member is not started yet.
+				log.Info("member is not started yet", zap.String("member-id", strconv.FormatUint(ep.GetID(), 16)), errs.ZapError(err))
+				continue
+			}
+			status, err := s.getClient().Status(ctx, ep.ClientURLs[0])
+			if err != nil {
+				log.Info("failed to get status of member", zap.String("member-id", strconv.FormatUint(ep.ID, 16)), zap.String("endpoint", ep.ClientURLs[0]), errs.ZapError(err))
+				continue
+			}
+			if status.Leader == ep.ID {
+				leaderAddr := ep.ClientURLs[0]
+				if s.pdLeaderAddr.CompareAndSwap(s.pdLeaderAddr.Load(), leaderAddr) {
+					if status.Leader != curLeader {
+						log.Info("switch PD leader", zap.String("leader-id", strconv.FormatUint(ep.ID, 16)), zap.String("endpoint", ep.ClientURLs[0]))
+						s.reconnectCh <- true
+					}
+					curLeader = ep.ID
+					break
+				}
+			}
+		}
+	}
+}
+
+func (s *RegionSyncer) triggerMembershipCheck() {
+	select {
+	case s.checkMembershipCh <- struct{}{}:
+	default: // avoid blocking
+	}
+}
+
+func (s *RegionSyncer) syncLoop() {
+	s.wg.Add(1)
```

Comment:
How about moving it before calling the function?

Reply:
done